flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferReaderWriterUtil.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+
+/**
+ * Putting and getting of a sequence of buffers to/from a FileChannel or a ByteBuffer.
+ * This class handles the headers, length encoding, memory slicing.
+ *
+ * <p>The encoding is the same across FileChannel and ByteBuffer, so this class can
+ * write to a file and read from the byte buffer that results from mapping this file to memory.
+ */
+final class BufferReaderWriterUtil {
+
+	static final int HEADER_LENGTH = 8;
+
+	static final int HEADER_VALUE_IS_BUFFER = 0;
+
+	static final int HEADER_VALUE_IS_EVENT = 1;
+
+	// ------------------------------------------------------------------------
+	//  ByteBuffer read / write
+	// ------------------------------------------------------------------------
+
+	static boolean writeBuffer(Buffer buffer, ByteBuffer memory) {
+		final int bufferSize = buffer.getSize();
+
+		if (memory.remaining() < bufferSize + HEADER_LENGTH) {
+			return false;
+		}
+
+		memory.putInt(buffer.isBuffer() ? HEADER_VALUE_IS_BUFFER : HEADER_VALUE_IS_EVENT);
+		memory.putInt(bufferSize);
+		memory.put(buffer.getNioBufferReadable());
+		return true;
+	}
+
+	@Nullable
+	static Buffer sliceNextBuffer(ByteBuffer memory) {
+		final int remaining = memory.remaining();
+
+		// we only check the correct case where data is exhausted
+		// all other cases can only occur if our write logic is wrong and will already throw
+		// buffer underflow exceptions which will cause the read to fail.
+		if (remaining == 0) {
+			return null;
+		}
+
+		final int header = memory.getInt();
+		final int size = memory.getInt();
+
+		memory.limit(memory.position() + size);
+		ByteBuffer buf = memory.slice();
+		memory.position(memory.limit());
+		memory.limit(memory.capacity());
+
+		MemorySegment memorySegment = MemorySegmentFactory.wrapOffHeapMemory(buf);
+
+		return bufferFromMemorySegment(
+				memorySegment,
+				FreeingBufferRecycler.INSTANCE,
+				size,
+				header == HEADER_VALUE_IS_EVENT);
+	}
+
+	// ------------------------------------------------------------------------
+	//  ByteChannel read / write
+	// ------------------------------------------------------------------------
+
+	static long writeToByteChannel(
+			FileChannel channel,
+			Buffer buffer,
+			ByteBuffer[] arrayWithHeaderBuffer) throws IOException {
+
+		final ByteBuffer headerBuffer = arrayWithHeaderBuffer[0];
+		headerBuffer.clear();
+		headerBuffer.putInt(buffer.isBuffer() ? HEADER_VALUE_IS_BUFFER : HEADER_VALUE_IS_EVENT);
+		headerBuffer.putInt(buffer.getSize());
+		headerBuffer.flip();
+
+		arrayWithHeaderBuffer[1] = buffer.getNioBufferReadable();
+
+		return channel.write(arrayWithHeaderBuffer);
```

Comment:
I am not sure if this function call can guarantee to write all data out, at least the java doc of GatheringByteChannel#write says that it may return zero. Maybe it is more safe to do the write in a while loop just like read.

Reply:
I was looking into this as well while writing. I think I found the docs saying that the file channel implementation does guarantee this, but I can double check and add a comment.

Reply:
I am adding a safety net. Seems cheap enough, and will put us on the safe side.